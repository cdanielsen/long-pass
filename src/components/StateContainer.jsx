import React from 'react'
import words from 'an-array-of-english-words'

import Header from './Header'
import SplashImage from './SplashImage'
import Grid from 'material-ui/Grid'
import SubmitButton from './SubmitButton'
import PasswordPresenter from './PasswordPresenter'
import WordLengthField from './WordLengthField'

// import PasswordGeneratorContainer from './PasswordGeneratorContainer'

class StateContainer extends React.Component {
  constructor () {
    super()
    this.state = {passwordContent: '', passwordMaxWordLength: 5}
    this.handleSubmitButtonClick = this.handleSubmitButtonClick.bind(this)
    this.handleMaxLengthInput = this.handleMaxLengthInput.bind(this)
  }

  pickRandomValue (array) {
    return (Math.floor(Math.random() * array.length))
  }

  generatePassword () {
    let newPass = []
    const maxPassLength = this.state.passwordMaxWordLength
    const filteredWords = words.filter(word => word.length <= maxPassLength)
    for (let i = 0; i < 4; i++) {
      newPass[i] = filteredWords[this.pickRandomValue(filteredWords)]
    }
    return newPass.join(' ')
  }

  handleSubmitButtonClick () {
    const newPass = this.generatePassword()
    this.setState((prevState, props) => (
      {passwordContent: newPass}
    ))
  }

  handleMaxLengthInput (event) {
    const maxLength = parseInt(event.target.value, 10)
    this.setState((prevState, props) => (
      {passwordMaxWordLength: maxLength}
    ), this.handleSubmitButtonClick)
  }

  render () {
    return (
      <div>
        <Grid container>
          <Grid
            item
            md={12}>
            <div>
              <Header
                id="banner"
                textContent='You Need A Long Password'/>
              <SplashImage
                id="main-image"
                imageUrl='https://imgs.xkcd.com/comics/password_strength.png'/>
              <SubmitButton
                id="new-password-submit"
                clickHandler={this.handleSubmitButtonClick}/>
            </div>
          </Grid>
        </Grid>
        <Grid container spacing={8}>
          <Grid
            item
            id="control-panel"
            md={6}>
            <div>
              <WordLengthField
                newSelectionHandler={this.handleMaxLengthInput}
                value={this.state.passwordMaxWordLength}/>
            </div>
          </Grid>
          <Grid
            item
            md={6}>

            <PasswordPresenter
              passwordContent={this.state.passwordContent}/>
          </Grid>
        </Grid>
      </div>
    )
  }
}

export default StateContainer
