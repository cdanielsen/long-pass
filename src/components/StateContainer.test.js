import React from 'react'
import StateContainer from './StateContainer.jsx'
import wordSet from 'more-words'
import { shallow } from 'enzyme'

describe('StateContainer render', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(<StateContainer />)
  })

  test('should render the correct tree of components', () => {
    expect(wrapper).toMatchSnapshot()
  })
})

describe('StateContainer helper methods', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(<StateContainer />)
  })

  describe('#pickRandomIndex', () => {
    test('should return a random index given an array', () => {
      const sampleArray = [1, 2, 3, 4, 5, 6, 7]
      const randomIndex = wrapper.instance().pickRandomIndex(sampleArray)
      expect(Number.isInteger(randomIndex)).toBe(true)
      expect(sampleArray.includes(sampleArray[randomIndex])).toBe(true)
    })
  })

  describe('#generatePassword', () => {
    test('should return an array', () => {
      const returnValueisArray = Array.isArray(
        wrapper.instance().generatePassword(wordSet)
      )
      expect(returnValueisArray).toBe(true)
    })

    test('should only generate values from the given dictionary', () => {
      const result = wrapper.instance().generatePassword(wordSet)
      const allWordsWereFromOriginalSet = result
        .map(word => wordSet.includes(word))
        .every(bool => bool === true)
      expect(allWordsWereFromOriginalSet).toBe(true)
    })
  })

  describe('#filterWordSet', () => {
    test('given a word length limit and a set of words, should filter out words over that limit from that set', () => {
      const state = wrapper.state()
      const defaultLengthLimit = state.passwordMaxWordLengthCurrent
      const result = wrapper.instance().generatePassword(wordSet)
      const resultContainsValuesLongerThanDefaultLimit = result.some(
        word => word.length > defaultLengthLimit
      )
      expect(resultContainsValuesLongerThanDefaultLimit).toBe(false)
    })
  })
})

describe('StateContainer event handlers', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(<StateContainer />)
  })

  describe('#handleSubmitButtonClick', () => {
    test('should set a new password value', () => {
      const initialPasswordContent = wrapper.state('passwordContent')
      expect(initialPasswordContent.length).toBe(0)
      wrapper.instance().handleSubmitButtonClick()
      const afterClickPasswordContent = wrapper.state('passwordContent')
      expect(afterClickPasswordContent.length).toBe(4)
    })
  })

  describe('#handleMaxWordLengthInput', () => {
    test('should trigger the generation of a new password value with words of limited length', () => {
      const newWordMaxLength = '4'
      wrapper.instance().handleMaxWordLengthInput({target: {value: newWordMaxLength}})
      const afterChangePasswordContent = wrapper.state('passwordContent')
      const contentContainsValuesLongerThanRequestedLimit = afterChangePasswordContent.some(
        word => word.length > newWordMaxLength
      )
      expect(contentContainsValuesLongerThanRequestedLimit).toBe(false)
    })
  })

  describe('#handleMaxPasswordLengthInput', () => {
    test('should trigger the generation of a new password value less than the new total length limit', () => {
      const newPassMaxLength = '16'
      wrapper.instance().handleMaxPasswordLengthInput({target: {value: newPassMaxLength}})
      const afterChangePasswordContent = wrapper.state('passwordContent')
      const actualPasswordLength = afterChangePasswordContent.join('').length
      const setLimit = parseInt(newPassMaxLength, 10)
      expect(actualPasswordLength < setLimit).toBe(true)
    })
  })

  describe('#handleUpdatingPasswordLength', () => {
    test('should calculate and set the passwordMaxWordLengthPossible value based on the current passwordMaxLength value', () => {
      const newPassMaxLength = 15
      const expectedMaxWordLengthPossible = 3
      // simulate a new passwordMaxLength value
      wrapper.setState({passwordMaxLength: newPassMaxLength})
      wrapper.instance().handleUpdatingPasswordLength()
      const actualMaxWordLengthPossible = wrapper.state('passwordMaxWordLengthPossible')
      expect(actualMaxWordLengthPossible).toBe(expectedMaxWordLengthPossible)
    })
    test('should leave the passwordMaxWordLengthCurrent value unchanged if falls within the limit of the new passWordMaxLength constraint', () => {
      const newPassMaxLength = 19
      const expectedMaxWordLengthCurrent = wrapper.state('passwordMaxWordLengthCurrent')
      // simulate a new passwordMaxLength value
      wrapper.setState({passwordMaxLength: newPassMaxLength})
      wrapper.instance().handleUpdatingPasswordLength()
      const actualMaxWordLengthCurrent = wrapper.state('passwordMaxWordLengthCurrent')
      expect(actualMaxWordLengthCurrent).toBe(expectedMaxWordLengthCurrent)
    })
    test('should set a new passwordMaxWordLengthCurrent value if the old value could violate the new passwordMaxWordLength constraint', () => {
      const newPassMaxLength = 10
      const expectedMaxWordLengthCurrent = 2
      wrapper.setState({passwordMaxLength: newPassMaxLength})
      wrapper.instance().handleUpdatingPasswordLength()
      const actualMaxWordLengthCurrent = wrapper.state('passwordMaxWordLengthCurrent')
      expect(actualMaxWordLengthCurrent).toBe(expectedMaxWordLengthCurrent)
    })
  })
})
