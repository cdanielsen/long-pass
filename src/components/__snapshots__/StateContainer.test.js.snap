// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateContainer render should render the correct tree of components 1`] = `
<div>
  <withStyles(GridWrapper)
    container={true}
  >
    <withStyles(GridWrapper)
      item={true}
      md={12}
    >
      <div>
        <Header
          className="banner"
          textContent="You Need A Long Password"
        />
        <SplashImage
          id="main-image"
          imageUrl="https://imgs.xkcd.com/comics/password_strength.png"
        />
        <Header
          className="banner"
          textContent="Make Me One!"
        />
        <SubmitButton
          clickHandler={[Function]}
          id="new-password-submit"
        />
      </div>
    </withStyles(GridWrapper)>
  </withStyles(GridWrapper)>
  <withStyles(GridWrapper)
    container={true}
    spacing={8}
  >
    <withStyles(GridWrapper)
      id="control-panel"
      item={true}
      md={6}
    >
      <div>
        <SliderFilter
          inputId="slider-max-password-length"
          max={28}
          min={12}
          newSelectionHandler={[Function]}
          title="Max Length"
          value={20}
        />
      </div>
    </withStyles(GridWrapper)>
    <withStyles(GridWrapper)
      id="control-panel"
      item={true}
      md={6}
    >
      <div>
        <SliderFilter
          inputId="slider-max-word-length"
          max={5}
          min={3}
          newSelectionHandler={[Function]}
          title="Max Word Length"
          value={3}
        />
      </div>
    </withStyles(GridWrapper)>
    <withStyles(GridWrapper)
      item={true}
      md={12}
    >
      <PasswordPresenter
        id="password-container"
        passwordContent={Array []}
      />
    </withStyles(GridWrapper)>
  </withStyles(GridWrapper)>
</div>
`;
